src/DynamicExpresso.Core/Parsing/Parser.cs

```diff
@@ -3346,6 +3350,9 @@ private void NextToken()
 					}
 					if (_parsePosition == _expressionTextLength)
 					{
+						if (_token.id == TokenId.End)
+							throw new InvalidOperationException("NextToken called when already at the end of the expression");
```

Comment:
I've added an exception if the parser tries to move past the end of the expression. The error in #287 would have been detected with this.